File: JavaScript - p5/Ship.js
const MAXSPEED = 15;

class Ship extends Entity {
    constructor(anim, x, y, heading) {
        super(anim, x, y, heading);
        this.thrusting = false;
        this.extraShips = 3;
    }
    update() {
        if (this.thrusting) {
            this.anim = animations["ship_thrusting"];
            sounds["thrust"].play();
            this.dx += Math.cos(this.heading) * 0.2;
            this.dy += Math.sin(this.heading) * 0.2;
        } else {
            this.anim = animations["ship"];
            sounds["thrust"].stop();
            this.dx *= 0.99;
            this.dy *= 0.99;
        }

        let speed = sqrt(this.dx * this.dx + this.dy * this.dy);
        if (speed > MAXSPEED) {
            this.dx *= MAXSPEED / speed;
            this.dy *= MAXSPEED / speed;
        }

        super.update();
        super.wrap();
    }
    checkCollisions() {
        for (let e of currentEntities) {
            if (e.collision(this)) {
                entities.push(new Explosion(animations["explosion_ship"], ship.x, ship.y));
                sounds["bang_lg"].play();
                this.x = width * 0.5; // reset the ship to the center of the screen
                this.y = height * 0.5;
                this.dx = this.dy = 0;

                if (--this.extraShips < 0) {
                    this.active = false;
                    if (saucer) saucer.destroy();
                    gameState = (score > lowestHighScore) ?
                        GameState.GAMEOVER : // prompt for player's name
                        GameState.READY; // show high scores
                }
            }
        }
    }
    drawExtraShips() {
        push();
        let f = animations["ship"].frames[0];
        scale(0.65);
        translate(1600, 100); // set the initial drawing position
        for (let i = 0; i < this.extraShips; i++) {
            translate(60, 0); // draw each ship to the left of the previous one
            image(animations["ship"].texture,
                0, 0, f.w, f.h,      // destination rectange -- where to draw the image
                f.x, f.y, f.w, f.h); // source of the image from within the larger texture
        }
        pop();
    }
}

File: Python - pygame/Ship.py
import pygame, math, Global
from Entity import *
from Explosion import *

class Ship(Entity):
  maxSpeed = 15

  def __init__(self, anim, x, y):
    super().__init__(anim, x, y, 0)
    self.thrusting = False
    self.thrustChannel = pygame.mixer.Channel(pygame.mixer.get_num_channels() - 1)

  def update(self):
    if self.thrusting:
      self.anim = Global.animations["ship_thrusting"]
      if not self.thrustChannel.get_busy():
        self.thrustChannel.play(Global.sounds["thrust"])
      self.dx += math.cos(math.radians(-self.heading)) * 0.2
      self.dy += math.sin(math.radians(-self.heading)) * 0.2
    else:
      self.anim = Global.animations["ship"]
      self.thrustChannel.stop()
      self.dx *= 0.99
      self.dy *= 0.99

    speed = math.sqrt(self.dx * self.dx + self.dy * self.dy)
    if speed > Ship.maxSpeed:
      self.dx *= Ship.maxSpeed / speed
      self.dy *= Ship.maxSpeed / speed

    super().update()
    super().wrap()  # wrap around the display edges

  def check_collisions(self):
    for e in Global.currentEntities:
      if (e.collides_with(self)):
        Explosion.spawn(Global.animations["explosion_ship"], self.x, self.y)
        Global.sounds["bang_lg"].play()
        self.x = Global.WIDTH * 0.5  # reset the Global.ship to the center of the screen
        self.y = Global.HEIGHT * 0.5
        self.dx = 0
        self.dy = 0
